[SelectionDAG][RISCV] Operations with static rounding

spavloff opened the pull request:
Some targets, including RISC-V, support rounding mode specified in instruction rather than in a register. To use this feature, new DAG nodes are introduced by this patch. They have the same operands as their default mode counterparts, but have additional integer parameter, which specifies the rounding mode to use. Lowering of these DAG nodes is implemented for extensions F and Zfinx.

spavloff committed:
[SelectionDAG][RISCV] Operations with static rounding

Some targets, including RISC-V, support rounding mode specified in
instruction rather than in a register. To use this feature, new DAG
nodes are introduced by this patch. They have the same operands as their
default mode counterparts, but have additional integer parameter, which
specifies the rounding mode to use. Lowering of these DAG nodes is
implemented for extensions F and Zfinx.

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -354,6 +354,16 @@ class TargetLoweringBase {
 
+  /// Returns true if the target supports static rounding mode for the given
+  /// instruction.
+  virtual bool isStaticRoundingSupportedFor(const Instruction &I) const {
+    return false;
+  }
```
This hook doesn't make sense to me, and I don't see how it would be usable

arsenm replied:
This hasn't been addressed. It makes little sense to query this from an IR instruction

spavloff replied:
This is default implementation, RISCV has its own. If you mean `getMachineRoundingMode`, which the first comment refered to, it is removed.

spavloff replied:
This callback is used by target-neutral SelectionDAGBuilder to decide how to lower a gived constrained intrinsic call, to STRICT_FADD, or to FADD_ROUND.

spavloff replied:
Shouldn't DAGBuilder just use the best target-specific implementation of a target-independent operation? The call: 

> call float @llvm.experimental.constrained.fadd.f32(float %x, float %y, metadata !"round.tonearest", metadata !"fpexcept.ignore")

is lowered to STRICT_FADD on X86 because there is no static rounding on this platform and DAG node anyway would be an operation with rounding mode taken from a hardware register. On RISCV, a better representation exist, so SelectionDAGBuilbder can use it.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8131,15 +8131,73 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    }
+    if (Opcode) {
+      int MachineRM = TLI.getMachineRoundingMode(*RM);
+      assert(MachineRM >= 0 && "Unsupported rounding mode");
```
Just pass it through and let it error later if need be

arsenm commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21427,6 +21432,23 @@ bool RISCVTargetLowering::preferScalarizeSplat(SDNode *N) const {
+int RISCVTargetLowering::getMachineRoundingMode(RoundingMode RM) const {
+  switch (RM) {
+  case RoundingMode::TowardZero:
+    return RISCVFPRndMode::RTZ;
```
Instead of introducing this hook, can these just be handled in selection values? It's only an encoding value question

spavloff replied:
It now is implemented using special node for RISCV, which differ from the target-neutral FADD_ROUND an other by using target dependent encoding. Selection DAG first builds target-neutral nodes, then they are loweded to RISCV-specific nodes, which in turn are lowered by TableGen patterns.

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8131,15 +8131,73 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    Flags.copyFMF(*FPOp);
+
+  bool UseStaticRounding = EB == fp::ExceptionBehavior::ebIgnore && RM &&
+                           *RM != RoundingMode::Dynamic &&
```
Isn't the contract for the constrained intrinsics, that the global rounding mode has already been changed? What do we gain by using the static rounding mode instructions?

If the global rounding mode hasn't already been changed, that means that whoever created the constrained intrinsics needs to know what operations support a static rounding mode in the target.

arsenm replied:
Yes. These would be useful if they were connected intrinsics with a fixed rounding mode (like we have fptrunc.round)

efriedma-quic replied:
Right... if we want floating-point ops with an explicitly specified rounding mode, we need IR support.  And "STRICT_" forms of the SelectionDAG ops.  If you want to go this route, it needs an RFC.

(If you just want the nodes for SelectionDAG-internal optimizations, you can skip this, though.)

spavloff replied:
The forthcoming C standard introduces static rounding mode in addition to the dynamic one (set by `fesetround`): https://www.iso-9899.info/n3047.html#7.6.2 . This rounding is set by pragma STDC FENV_ROUND and may be different from the dynamic rounding. This model well fits architecture of targets, where rounding mode can be encoded in an instruction. RISCV is such in the LLVM tree, but there are other processors (maybe most AI cores) that support this feature. There is a demand to support static rounding mode in LLVM.

Constrained intrinsics are not the best representation for static rounding mode in IR, because they have side effect, but the can be used as a starting point. In future new ways to express associated rounding mode can appear, may be as proposed in https://discourse.llvm.org/t/thought-on-strictfp-support/71453.

arsenm replied:
But that's not what you are implementing here. You are implementing operations that do read the dynamic mode into ones that don't. We would need a separate set of operations corresponding to cases with a static mode. The metadata is not a request to use a specific mode, they only indicate what the mode is.

> Constrained intrinsics are not the best representation for static rounding mode in IR

These aren't a representation of it at all, it's conceptually different.

spavloff replied:
What is wrong with using constrained intrinsic for operations with static rounding?

LangRef defines these intrinsics (https://llvm.org/docs/LangRef.html#constrained-floating-point-intrinsics):

> These intrinsics are used to provide special handling of floating-point operations when specific rounding mode or floating-point exception behavior is required. 

This is just what is required for static rounding operations. The bad thing is the side effect, which is present on every constrained intrinsic. It however does not break the semantics of static rounding, it only prevents some optimization.

Constrained intrinsics are used here as per their desigs, no?

arsenm replied:
> What is wrong with using constrained intrinsic for operations with static rounding?

It's not really what they are for. This isn't wrong, but why would using the fixed mode instructions be better than using instructions that do read the mode? The constrained argument only tells you the dynamic mode will be a certain value. Using fixed mode instructions does not change the expected behavior. Is there some reason using the fixed mode instructions is an optimization (I'd assume not)?

> Constrained intrinsics are used here as per their desigs, no?

Not really. What you're describing is more like https://llvm.org/docs/LangRef.html#llvm-fptrunc-round-intrinsic, where the rounding mode argument provides the behavior to use. The backend must manage mode switches as necessary, or use fixed rounding mode instructions like these ones you are defining.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.h:
```diff
@@ -535,6 +535,11 @@ class RISCVTargetLowering : public TargetLowering {
   bool softPromoteHalfType() const override { return true; }
 
+  bool isStaticRoundingSupportedFor(const Instruction &I) const override {
+    return true;
```
This is wrong for vectors.

spavloff committed:
Address review comments

- change node names from *_MODE to *_ROUND. Also rename related objects,
- remove handling fptrunc_round. Tests for it require another FP type,
  which suport static rounding.
- get rid of the target callback getMachineRoundingMode. It requires new
  RISCV-specific nodes to represent the nodes with machine-specific
  rounding mode,
- provide implementaton of isStaticRoundingSupportedFor.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8131,15 +8131,69 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    Opers.push_back(getValue(FPI.getArgOperand(I)));
+
+  if (UseStaticRounding) {
+    switch (FPI.getIntrinsicID()) {
```
It still doesn't make sense to me to select the constrained intrinsics to these operations. The metadata indicates that the runtime mode is this already, there isn't a benefit to using an instruction with a fixed mode

spavloff replied:
Nodes like `STRICT_FADD` do not have information about rounding mode, they only prevent from undesired reordering by having side effect. After selector the information about rounding mode disappears.

spavloff committed:
Add chech for F/Zfinx
